### PR TITLE
Split out VerifiableMap from SparseMerkleTree.

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: cpp/merkletree/sparse_merkle_tree.h [hunk 1]
```diff
@@ -260,4 +238,19 @@ inline int PathBit(const SparseMerkleTree::Path& path, size_t bit) {
   return (path[bit / 8] & (1 << (7 - bit % 8))) == 0 ? 0 : 1;
 }
 
+
+namespace std {
+
+
+template <>
+struct hash<SparseMerkleTree::Path> {
```
Comment (pphaneuf) [inline cpp/merkletree/sparse_merkle_tree.h]: https://google.github.io/styleguide/cppguide.html#std_hash

File: cpp/merkletree/sparse_merkle_tree.cc [hunk 2]
```diff
@@ -64,16 +58,16 @@ void SparseMerkleTree::SetLeaf(const Path& path, const string& data) {
...
-                .emplace(make_pair(node_index, TreeNode(path, data)))
+                .emplace(make_pair(node_index, TreeNode(path, leaf_hash)))
                 .second);
       return;
     } else if (it->second.type_ == TreeNode::INTERNAL) {
       // Mark the internal node hash dirty
-      it->second.hash_.reset();
-    } else if (it->second.leaf_->path_ == path) {
+      it->second.hash_.clear();
+    } else if (*(it->second.path_) == path) {
```
Comment (pphaneuf) [inline cpp/merkletree/sparse_merkle_tree.cc]: Those parens are needed?

**Comment by pphaneuf** [comment]
Done with the review, just make the hash explicit (instead of specializing `std::hash`), and we'll be on our way!

File: cpp/merkletree/sparse_merkle_tree.h (hunk 1 above)
Comment (AlCutter) [inline cpp/merkletree/sparse_merkle_tree.h]: hehe, ok - made explicit.

File: cpp/merkletree/sparse_merkle_tree.cc (hunk 2 above)
Comment (AlCutter) [inline cpp/merkletree/sparse_merkle_tree.cc]: actually nope, removed.

File: cpp/merkletree/sparse_merkle_tree.h [hunk 3]
```diff
@@ -260,4 +239,13 @@ inline int PathBit(const SparseMerkleTree::Path& path, size_t bit) {
   return (path[bit / 8] & (1 << (7 - bit % 8))) == 0 ? 0 : 1;
 }
 
+
+struct PathHasher {
+  size_t operator()(const SparseMerkleTree::Path& p) const {
+    return std::hash<std::string>()(
+        std::string(reinterpret_cast<const char*>(p.data()), p.size()));
```
Comment (pphaneuf) [inline cpp/merkletree/sparse_merkle_tree.h]: Could ding the `const`.

**Comment by pphaneuf** [comment]
One nit, but LGTM otherwise.

File: cpp/merkletree/sparse_merkle_tree.h (hunk 3 above)
Comment (AlCutter) [inline cpp/merkletree/sparse_merkle_tree.h]: Dinged

**Pull request merged**